#My script for using DESEQ2
# ### From the Pipeline script
# ### it's looking at the variables given to it by the submission script
# if [[ "$Rdeseq" == "yes" ]]; then ## if I want to use DESeq, which I do
# 
# for file in $deseqFinalProcessR $dataframe; do  ### 1 is the DESeq R script, 2 is the sample information file
# if [ ! -e $file ]; then echo "$file does not exist"; exit; fi #check that these both exist
# done
# 
# #### add this to the main script that's being generated by the pipeline
# echo "
# ${Rbin} CMD BATCH --no-save --no-restore --keep.sex=${keepSex} --support.frame=${dataframe} --keep.dups=${keepDups} --code=${code} --annotation.file=${annotationFile} --iFolder=${oFolder} ${deseqFinalProcessR} ${clusterFolder}/R/deseq_${stem}.out 
# " >> $mainscript
# #### dataframe is the support file, keepDups is whether to keep duplicates or not, code is the name given to the job (eg Jack_Job), annotation.file is a biomart human annotations file, stem is same as code unless specified otherwise
# 

library(DESeq2)
### create a function that reads in the list of arguments given by the pipeline script
getArgs <- function() {
  myargs.list <- strsplit(grep("=",gsub("--","",commandArgs()),value=TRUE),"=")  ## commandArgs is the list of command arguments given to the script, this part of the function splits off the -- part
  myargs <- lapply(myargs.list,function(x) x[2] )
  names(myargs) <- lapply(myargs.list,function(x) x[1])
  return (myargs) + 
}

### Just for debugging 
support.frame <- "/cluster/project8/vyp/Tabrizi_Huntington_RNASeq/support/htt_support2.txt"
code <- "htt"
keep.dups <- FALSE 
annotation.file <- "/cluster/project8/vyp/vincent/Software/RNASeq_pipeline/bundle/human/biomart/biomart_annotations_human.tab"
iFolder <- "/scratch2/vyp-scratch2/Tabrizi_Huntington_RNASeq/processed/Nov2014/"

#more debugging - Pietro Adult F210I
support.frame <- "project/Humphrey_C9orf72_longGenes/long_genes/fratta_adult_f210i/design_jack.tab"
counts <- "project/Humphrey_C9orf72_longGenes/long_genes/fratta_adult_f210i/counts_Fratta_deseq_WTvsHE.RData"
annotation.file <- "project/vincent/Software/RNASeq_pipeline/bundle/mouse/biomart/biomart_annotations_mouse.tab"
#clean counts from SVA
counts <- "/Users/Jack/long_genes/fratta_adult_f210i/counts_fratta_f210i_adult_2_sva_clean.RData"
#set all adjusted count values as integers
load(counts)
genes.counts <- apply(X=clean.genes.counts, MARGIN = c(1,2),FUN = function(x) round(x,digits=0))

## vincent debugging
support.frame <- "data/TDP43_m323k.tab"
code <- "m323k"
annotation.file <- "/cluster/project8/vyp/vincent/Software/RNASeq_pipeline/bundle/mouse/biomart/biomart_annotations_mouse.tab"
iFolder <- "/scratch2/vyp-scratch2/IoN_RNASeq/Fratta_RNASeq/brain/m323k"
keep.dups <- FALSE
keep.sex <- FALSE

########################## read arguments using the getArgs function

myArgs <- getArgs()
if ('support.frame' %in% names(myArgs)) support.frame <- myArgs[['support.frame']]
#added a counts argument which specifies where the counts table is
if ('counts' %in% names(myArgs)) counts <- myArgs[['counts']]
if ('code' %in% names(myArgs)) code <- myArgs[['code']]
if ('iFolder' %in% names(myArgs)) iFolder <- myArgs[['iFolder']]
if ('annotation.file' %in% names(myArgs)) annotation.file <- myArgs[['annotation.file']]
if ('keep.dups' %in% names(myArgs)) keep.dups <- as.logical(myArgs[['keep.dups']])
if ('keep.sex' %in% names(myArgs)) keep.sex <- as.logical(myArgs[['keep.sex']]) 

extra.plots <- TRUE 
remove.hb   <- FALSE 

message("Should I keep the sex chromosomes? Option set is ", keep.sex)

### FOR TESTING PURPOSES
#support.frame = "project/Humphrey_C9orf72_longGenes/long_genes//isaacs_c9/design_c9_trimmed_case_control.tab"

##hard code annotation file for now
#annotation <- read.table("project/vincent/Software/RNASeq_pipeline/bundle/human/biomart/biomart_annotations_human.tab",header=T)

###check input files and data frame
message('Now reading ', support.frame)
## read in support table, create lists of conditions and covariates
support <- read.table(support.frame, header = TRUE, stringsAsFactors = FALSE) 
list.conditions <- grep(names(support), pattern = '^condition.*', value  = TRUE) #looks for what conditions are referred to, either "condition" or "conditions"
list.covars <- grep(names(support), pattern = '^covar.*', value  = TRUE)

## read in annotation file

annotation <- read.table(annotation.file, header = TRUE, sep = '\t', na.string = c('', 'NA'), quote = "" )
names(annotation) <- ifelse (names(annotation) == "external_gene_name", "external_gene_id", names(annotation)) # trying to agree on the column names

print(head(annotation))

### create deseq output folders and files
deseq2.folder <- paste(iFolder, '/deseq2', sep = '')
for (folder in c(deseq2.folder)) {
  if (! file.exists(folder)) dir.create(folder)
}


########## load the count data

# if (keep.dups) deseq.counts <- paste(deseq2.folder, '/deseq_counts_', code, '_keep_dups.RData', sep = '')  
# if (!keep.dups) deseq.counts <- paste(deseq2.folder, '/deseq_counts_', code, '.RData', sep = '')  
# load(deseq.counts)

### hard code in the previously generated counts file - Change me!
#deseq.counts <- "project/Humphrey_C9orf72_longGenes/c9_qc/processed/deseq2/deseq_counts_Humphreys_C9orf47_keep_dups.RData"

deseq.counts <- counts
load(deseq.counts)

### Remove the sex chromosome genes 
if (!keep.sex) {
  genes.on.XY <- as.character(subset(annotation, chromosome_name %in% c('X' ,'Y'), 'EnsemblID', drop = TRUE))
  message('Prior to removing chr XY probes: ', nrow(genes.counts))
  genes.counts <- genes.counts[ ! dimnames(genes.counts)[[1]] %in% genes.on.XY, ]                 
  message('After removing chr XY probes: ', nrow(genes.counts))
}


### Remove the hemoglobin genes (for whole blood only) 
if(remove.hb) { 
   message("Removing HBB, HBA1 and HBA2") 
   hb.genes <- annotation[which(annotation$external_gene_id %in% c("HBB", "HBA1", "HBA2")), 'EnsemblID'] 
   genes.counts <- genes.counts[ ! dimnames(genes.counts)[[1]] %in% hb.genes, ]                 
   message('After removing hb genes: ', nrow(genes.counts))
} 


### Loop over all proposed conditions
for (condition in list.conditions) {
  num.cond <- FALSE 
  message("Processing for ", condition) 
  samples.to.use   <- !is.na(support[,condition]) #creates a vector of TRUE/FALSE values to use to filter out samples you're not interested in comparing
  
  genes.counts.loc <- genes.counts[, samples.to.use ] #from gene counts RData, take only the columns of the samples you're interested in

  support.loc <-  support[  samples.to.use, ] #extracts only the relevent samples from the support file

## determines if the condition be treated as a factor or a numeric variable 
  if (substr(condition, 10, 12) == "Num") { 
     num.cond <- TRUE 
     support.loc$condition <- as.numeric(support.loc[, condition])
     loc.code <-  condition
  } else { 
     support.loc$condition <- factor(support.loc[, condition]) #ensures that each condition is treated as a factor
     loc.code <-  paste(unique(support.loc$condition), collapse = '_') #creates a code "condition1_condition2". This will be in the alphabetical order of the conditions
  } 

################### create the appropriate folders and specify output file
  
loc.deseq2.folder <- paste(iFolder, code, sep = '') #make a deseq folder within the ifolder
  deseq2.figs <- paste(loc.deseq2.folder, '/figs', sep = '') #make a figs folder within the deseq folder

  for (folder in c(loc.deseq2.folder, deseq2.figs)) {
    if (! file.exists(folder)) dir.create(folder)
  }
  
  if (keep.dups) output.file <- paste(loc.deseq2.folder, '/deseq_', code, '_differential_expression_keep_dups.tab', sep = '')
  if (!keep.dups) output.file <- paste(loc.deseq2.folder, '/deseq_', code, '_differential_expression.tab', sep = '')

###################
  use.covars <- FALSE
  if (length(list.covars) > 0) {
     use.covars <- TRUE
  }
  
  if (use.covars) {
    message("Using ", length(list.covars), " covariates") 
    formula1 <- paste(" ~ ", paste(list.covars, collapse = "+"), " + ", condition, sep = "") 
    formula1 <- as.formula(formula1) 
    formula0 <- paste(" ~ ", paste(list.covars, collapse = "+"), sep = "") 
    formula0 <- as.formula(formula0) 
   
    design.deseq <- support.loc[, which(names(support.loc) %in% c(condition, list.covars))]
  } else {
    formula1 <- as.formula(paste0("~ ", condition))
    formula0 <- as.formula("~ 1")
    design.deseq <- support.loc[, c(condition), drop = FALSE] #take just the condition column of the relevent samples from the support file
  }
  head(genes.counts.loc)
  print(design.deseq)
  
  CDS <- DESeqDataSetFromMatrix(countData = genes.counts.loc, colData = design.deseq, design = formula1) #create the DESeqDataSet from the counts tables according to the conditions in the design table and according to the formula.

#################### Do the actual model fitting 
  CDS <- DESeq(CDS, test = "LRT", reduced = formula0, 
               minReplicatesForReplace = 5 ) 
  deseq.res <- results(CDS)  ## extract the results from the DESEq table

  ### output the design information
  row.names(design.deseq) <- support.loc$sample ## make the row names of design.deseq the name of each sample
  write.table(x = design.deseq, file = paste(loc.deseq2.folder, '/design.tab', sep = ''), row.names = TRUE, quote = FALSE, sep = '\t') #output the design.deseq as a conditions output
  
############# Make the results table into a sensible format 
  deseq.res.df <- data.frame(deseq.res) ## convert the results table into a data frame 
  print(head(deseq.res.df)) 
  deseq.res.df$EnsemblID <- row.names( deseq.res.df) #add EnsemblID column by transferring across the row names
  deseq.res.df <- merge(deseq.res.df, annotation, by = 'EnsemblID', all.x = TRUE) #merge with annotation table to give gene name and length info
  deseq.res.df <- deseq.res.df[order(deseq.res.df$pvalue),] #sort by p value
  print(head(deseq.res.df)) 

#################### now write the output to a file 
  write.table(deseq.res.df, file = output.file, quote = FALSE, row.names = FALSE, sep = "\t" ) 
loc.code <- code
  #save(CDS, file = "CDS.RData")
######### Now add a PCA for the subset of individuals being considered
#   if(extra.plots) { 
#      if (keep.dups) { 
#         output.pca <- paste(deseq2.figs, '/', loc.code, '_pca_keepdups.pdf', sep = '')
#      } else { 
#         output.pca <- paste(deseq2.figs, '/', loc.code, '_pca.pdf', sep = '')
#      } 
#      pdf(output.pca) 
#      DESeq2::plotPCA(CDS, intgroup = condition) 
#      dev.off() 
output.pca <- paste(deseq2.figs, '/', loc.code, '_pca.pdf', sep = '')

vsd <- varianceStabilizingTransformation(CDS)
source('/cluster/project8/vyp/Humphrey_C9orf72_longGenes/c9_deseq//tools.R')
source('project/Humphrey_C9orf72_longGenes/c9_deseq/tools.R')
myPCA(vsd,output.pca)

## Visualise the counts versus condition for the genes with best p-values 
     if (keep.dups) { 
        output.sig.genes <- paste(deseq2.figs, '/', loc.code, "_siggenes_keepdups.pdf", sep = '') 
     } else { 
        output.sig.genes <- paste(deseq2.figs, '/', loc.code, "_siggenes.pdf", sep = '') 
     }

     if (!num.cond) { 
     pdf(output.sig.genes) 
     sig.genes <- which(deseq.res.df$padj < 0.1)
     for (i in sig.genes) {
       plotCounts(CDS, gene = deseq.res.df$Ensembl[i], intgroup = condition,
                  transform = TRUE, main = deseq.res.df$external_gene_id[i])
       mtext(paste("pval: ", deseq.res.df$padj[i], " length: ", deseq.res.df$length[i], sep = ""), 3)
     }
     } 
     dev.off() 
     } 

     if (keep.dups) { 
        disp.plot <- paste(deseq2.figs, '/', loc.code, "_disp_keepdups.pdf", sep = '') 
     } else { 
        disp.plot <- paste(deseq2.figs, '/', loc.code, "_disp.pdf", sep = '') 
     } 

     pdf(disp.plot) 
     plotDispEsts(CDS)  
     dev.off() 

###GGplot for labelled counts

library(gridExtra)
library(ggplot2)
#based on this code
ggplot(data, aes(x=condition, y=count, color=sample)) +
  #scale_y_log10() + 
  geom_point(position=position_jitter(width=.1,height=0.5),size =5) +
  theme(title = deseq.res.df$external_gene_id[i])

pdf("long_genes/fratta_adult_f210i/fratta_f210i_adult_sva_2_top_20.pdf",onefile = T)
for (i in sig.genes) {
  data <- plotCounts(CDS, gene= deseq.res.df$Ensembl[i], intgroup=condition, returnData=TRUE)
  data$sample <- c("WT1","WT2","WT3","WT4","HET1","HET2","HET3","HET4")
  title = paste0(toupper(deseq.res.df$external_gene_id[i]),"\n P: ", round(deseq.res.df$padj[i],10),"\n Length: ", round(deseq.res.df$length[i] / 1e3), "kb")
  plot <- ggplot(data, aes(x=condition, y=count, color = sample)) + geom_point(position=position_jitter(width=.1,height=0.5),size=5) + ggtitle(title) + theme(plot.title = element_text(lineheight=.8, face="bold"))
  grid.arrange(plot)
} 
dev.off()
#alternate with WT3 removed
pdf("long_genes/fratta_adult_f210i/outlier_search_wt3_removed.pdf",onefile = T)
for (i in sig.genes) {
  data <- plotCounts(CDS_without_3, gene= new$Ensembl[i], intgroup=condition, returnData=TRUE)
  data$sample <- c("WT1","WT2","WT4","HET1","HET2","HET3","HET4")
  title = paste0(toupper(new$external_gene_id[i]),"\n P: ", round(new$pvalue[i],4),"\n Length: ", round(new$length[i] / 1e3), "kb")
  plot <- ggplot(data, aes(x=condition, y=count, color = sample)) + geom_point(position=position_jitter(width=.1,height=0.5),size=5) + ggtitle(title) + theme(plot.title = element_text(lineheight=.8, face="bold"))
  grid.arrange(plot)
} 
dev.off()
  
  ## create Z score vs length plots
data <- deseq.res.df
data$zscore <- qnorm(p = 1 - data$pvalue/2)
data$zscore<- ifelse ( data$log2FoldChange > 0, abs(data$zscore), - abs(data$zscore))
data$length <- data$end_position - data$start_position
#remove genes shorter than 1kb
#data <- subset(data,length > 1000)

data <- data[ order(data$zscore), ]
data$ranked.group <- rep(1:1000, each = 100)[ 1:nrow(data) ]
mean.length.by.group <- tapply(data$length, FUN = mean, INDEX = data$ranked.group, na.rm = TRUE)
average.z.score.by.group <- tapply(data$zscore, FUN = mean, INDEX = data$ranked.group, na.rm = TRUE)

z.plot <- paste(deseq2.figs, '/',loc.code, "_average_z_score_v_length.pdf", sep ='')
z.ranked.plot <- paste(deseq2.figs, '/', loc.code, "_ranked_z_score_v_length.pdf", sep ='')
z.length.plot <- paste(deseq2.figs, '/', loc.code, "_z_score_v_length.pdf", sep = '')

pdf(z.plot)
plot(x = average.z.score.by.group, y = mean.length.by.group, ylab = "mean length by group/bp", main = paste(loc.code, "Z score against length, grouped genes", sep = ' '))
dev.off()

pdf(z.ranked.plot)
plot(x = mean.length.by.group, xaxt = "n", xlab ="", ylab = "Mean gene length by group / base pairs", main = paste(loc.code, "grouped genes ranked by length", sep = ' '))
labels <- c("Most downregulated","Not regulated","Most upregulated")
mtext(labels, side = 1,line = 0.5,at = c(25,100,180))
dev.off()

pdf(z.length.plot)
plot(y = data$zscore, x = data$length,xlab = "gene length (bp)", ylab="Signed Z score", main = paste(loc.code, "Z score against length, all genes", sep = ' '))
dev.off()

  } # End of extra plots  

} # End looping over conditions 

warnings()


sessionInfo()

pdf("pairwiseMAs.pdf")
MA.idx = t(combn(1:8, 2))
for( i in  seq_along( MA.idx[,1])){ 
  MDPlot(counts(CDS, normalized = T)[idx.nz ,], 
         c(MA.idx[i,1],MA.idx[i,2]), 
         main = paste( colnames(CDS)[MA.idx[i,1]], " vs ",
                       colnames(CDS)[MA.idx[i,2]] ), ylim = c(-3,3))
}
dev.off()

dataGG = data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2], 
                    PC3 = PCA$x[,3], PC4 = PCA$x[,4], 
                    sampleNO = colData(rld)$sampleNO,
                    condition = colData(rld)$condition)

(qplot(PC1, PC2, data = dataGG, color =  condition, 
       main = "PC1 vs PC2, top variable genes", size = I(6))
 + labs(x = paste0("PC1, VarExp:", round(percentVar[1],4)),
        y = paste0("PC2, VarExp:", round(percentVar[2],4)))
 + scale_colour_brewer(type="qual", palette=2)
)

Pvars <- rowVars(assay(rld))
select <- order(Pvars, decreasing = TRUE)[seq_len(min(ntop, 
                                                      length(Pvars)))]

PCA <- prcomp(t(assay(rld)[select, ]), scale = F)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)

dataGG = data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2], 
                    PC3 = PCA$x[,3], PC4 = PCA$x[,4], 
                    sampleNO = colData(rld)$sampleNO,
                    condition = colData(rld)$condition)

(qplot(PC1, PC2, data = dataGG, color =  condition, 
       main = "Fratta Adult F210I PC1 vs PC2, top variable genes, no outliers", size = I(6))
 + labs(x = paste0("PC1, VarExp:", round(percentVar[1],4)),
        y = paste0("PC2, VarExp:", round(percentVar[2],4)))
 + scale_colour_brewer(type="qual", palette=2)
)
